memorystore/valkey/caching/GUIDE.md

```diff
@@ -0,0 +1,189 @@
+2. Select Sprint Boot version 3.4.1
+3. Complete the appropriate metadata.
+4. Choose your preffered Packing for downloading
+5. Select `Java 17` for your Java version.
+6. Finally, generate and extract the files.
+
+### Installing additional dependencies
+
+Next, ensure the following dependencies have been added to your POM.xml file.
+
+#### Jedis
+
+Add the folowing snippet toconnect directly to the Memorystore for Valkey instance.
+
+```xml
+        <!-- Jedis: Redis Java Client -->
+        <dependency>
+            <groupId>redis.clients</groupId>
+            <artifactId>jedis</artifactId>
+            <version>4.3.0</version> <!-- Use the latest version -->
+        </dependency>
+```
+
+#### Jakarta
+
+To ensure that our api routes are correctly validated. Add the following dependency:
+
+```xml
+   <!-- Add Validation support-->
+   <dependency>
+      <groupId>jakarta.validation</groupId>
+      <artifactId>jakarta.validation-api</artifactId>
+      <version>3.0.2</version>
+   </dependency>
+```
+
+### Connecting our Service layer
+
+Next, add the following to route logic to the API.
+
+### Writing to the cache
+
+Adding an entry will add the new item to the database. Once created, the id return from the database will be updated to a new object with the entries attributes. This new object will be added to the Memorystore cach with the appropirate Time-to-live (TTL) value.
+
+```java
+
+/** Import the Jedis library */
+import redis.clients.jedis.Jedis;
+
+/** Creating an item with Time-to-live (TTL) */
+public long create(Item item) {
+   /** Save the item in the database */
+   long itemId = itemsRepository.create(item);
+
+   /** Create a new object with the saved database id */
+   Item createdItem = new Item(
+   itemId,
+   item.getName(),
+   item.getDescription(),
+   item.getPrice()
+   );
+
+   /** Cache the data in Memorystore for valkey with the Time-to-live value **/
+   String idString = Long.toString(itemId);
+   jedis.set(idString, createdItem.toJSONObject().toString());
+   jedis.expire(idString, DEFAULT_TTL);
+
+   /** Return the item id */
+   return itemId;
+}
+```
+
+### Reading from the Cache (Retrieving Values)
+
+This method demonstrates how to efficiently retrieve data using a caching layer (Memorystore) to improve performance. The function first checks the cache for the requested item and handles its time-to-live (TTL) appropriately. If the item is not found in the cache, it fetches the data from the database, caches it, and then returns the result.
+
+```java
+/** Import the Jedis library */
+import redis.clients.jedis.Jedis;
+
+public Item get(long id) {
+   /** Ensure that the item id is a string for retirval from Memorystore */
+    String idString = Long.toString(id);
+
+    /* Check if the item exists in  the cache */
+    if (jedis.exists(idString)) {
+      // If the data exists in the cache extend the TTL
+      jedis.expire(idString, DEFAULT_TTL);
```

Comment:
i'm curious what the purpose of this is. If the purpose is to do some sort of least-recently-used eviction algorithm why not just use redis built in allkeys-lru maxmemory-policy?

The reason I ask is now we're converting every read into 2 reads and a write which is going to triple QPS.

also TTL often indicates some sort of allowed staleness. this technique will allow a frequently read item to stay in cache indefinitely - allowing it to get very stale.

Reply:
This example has been updated, with new instructions prior explaining the meaning behind the code example.